add async_http_retry_request_middleware, add tests for async_http_ret…

LaoGunz opened the pull request:
there was no method for http_retry_request_middleware for AsyncWeb3

What was wrong?
closes #3003 

How was it fixed?
add async_http_retry_request_middleware, add tests for async_http_ret…

Todo:
- [x] Document async_http_retry_request_middleware
- [x] Add relevant newsfragments

![](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2F4.bp.blogspot.com%2F-kRzR25oOHS4%2FV9Bcm6tC7PI%2FAAAAAAAAQr4%2FtAcZ2FXjrusSHeyTjMQo8cEO-WiHUuR6QCLcB%2Fs1600%2FPangolin.jpg&f=1&nofb=1&ipt=6863178b6687355e559d4433f8c1e1e89278331478041335e57aa337bb5ac79a&ipo=images)

ArtemSmeta committed:
add async_http_retry_request_middleware, add tests for async_http_retry_request_middleware

fselmo commented:
This looks great @LaoGunz. I'm going to add relevant newsfragments and docs and merge. Thanks for submitting 👍🏼

fselmo committed:
Changes to ``async_http_retry_request_middleware`` before merging

- ``ClientTimeout`` is not an exception but a timeout class.
- Catch more common and more broad exception classes as the defaults.
- More robust tests

fselmo committed:
Update docs for PR #3009

fselmo committed:
Make async_http_retry_request_middleware default for AsyncHTTPProvider

fselmo committed:
Update tests to match default settings

fselmo committed:
Add async_http_retry_request_middleware to middleware __init__.py

kclowes reviewed (approved):
LGTM! 🚢
